backends/tfhe-cuda-backend/cuda/src/integer/scalar_shifts.cuh

```diff
@@ -107,4 +109,127 @@ __host__ void host_integer_radix_scalar_shift_kb_inplace(
   }
 }
 
+template <typename Torus>
+__host__ void scratch_cuda_integer_radix_arithmetic_scalar_shift_kb(
+    cuda_stream_t *stream, int_arithmetic_scalar_shift_buffer<Torus> **mem_ptr,
+    uint32_t num_radix_blocks, int_radix_params params, SHIFT_TYPE shift_type,
+    bool allocate_gpu_memory) {
+
+  cudaSetDevice(stream->gpu_index);
+  *mem_ptr = new int_arithmetic_scalar_shift_buffer<Torus>(
+      stream, shift_type, params, num_radix_blocks, allocate_gpu_memory);
+}
+
+template <typename Torus>
+__host__ void host_integer_radix_arithmetic_scalar_shift_kb_inplace(
+    cuda_stream_t *stream, Torus *lwe_array, uint32_t shift,
+    int_arithmetic_scalar_shift_buffer<Torus> *mem, void *bsk, Torus *ksk,
+    uint32_t num_blocks) {
+
+  cudaSetDevice(stream->gpu_index);
+  auto params = mem->params;
+  auto glwe_dimension = params.glwe_dimension;
+  auto polynomial_size = params.polynomial_size;
+  auto message_modulus = params.message_modulus;
+
+  size_t big_lwe_size = glwe_dimension * polynomial_size + 1;
+  size_t big_lwe_size_bytes = big_lwe_size * sizeof(Torus);
+
+  size_t num_bits_in_block = (size_t)log2(message_modulus);
+  size_t total_num_bits = num_bits_in_block * num_blocks;
+  shift = shift % total_num_bits;
+
+  if (shift == 0) {
+    return;
+  }
+  size_t rotations = std::min(shift / num_bits_in_block, (size_t)num_blocks);
+  size_t shift_within_block = shift % num_bits_in_block;
+
+  Torus *rotated_buffer = mem->tmp_rotated;
+  Torus *padding_block = &rotated_buffer[num_blocks * big_lwe_size];
+  Torus *last_block_copy = &padding_block[big_lwe_size];
+
+  auto lut_univariate_shift_last_block =
+      mem->lut_buffers_univariate[shift_within_block - 1];
+  auto lut_univariate_padding_block =
+      mem->lut_buffers_univariate[num_bits_in_block - 1];
+  auto lut_bivariate = mem->lut_buffers_bivariate[shift_within_block - 1];
+
+  if (mem->shift_type == RIGHT_SHIFT) {
+    radix_blocks_rotate_left<<<num_blocks, 256, 0, stream->stream>>>(
+        rotated_buffer, lwe_array, rotations, num_blocks, big_lwe_size);
+    cuda_memcpy_async_gpu_to_gpu(lwe_array, rotated_buffer,
+                                 num_blocks * big_lwe_size_bytes, stream);
+    cuda_synchronize_stream(stream);
```

Comment:
I don't think we need this sync.

Reply:
Ah I reproduced what 's done in the cmux, where omp is used as well actually.

Reply:
It is actually necessary, otherwise the test fails.